lua/mason-null-ls/init.lua

```diff
@@ -1,140 +1,135 @@
+		Optional.of_nilable(handlers[source_name]):or_(_.always(default_handler)):if_present(function(handler)
+			log.fmt_trace('Calling handler for %s', source_name)
+			local ok, err = pcall(handler, source_name)
+			if not ok then
+				vim.notify(err, vim.log.levels.ERROR)
+			end
+		end)
+	end
 
-local show = function(msg)
-	vim.schedule_wrap(print(string.format('[mason-null-ls] %s', msg)))
+	local installed_sources = _.filter_map(get_source_name, registry.get_installed_package_names())
+	_.each(call_handler, installed_sources)
+	registry:on(
+		'package:install:success',
+		vim.schedule_wrap(function(pkg)
+			get_source_name(pkg.name):if_present(call_handler)
+		end)
+	)
 end
 
-local show_error = function(msg)
-	vim.schedule_wrap(vim.api.nvim_err_writeln(string.format('[mason-null-ls] %s', msg)))
-end
+---@return string[]
+function M.get_installed_sources()
+	local Optional = require('mason-core.optional')
+	local registry = require('mason-registry')
+	local source_mapping = require('mason-null-ls.mappings.source')
 
-local function auto_get_packages()
-	local sources = {}
-	sources = vim.tbl_deep_extend('force', sources, vim.tbl_keys(require('null-ls.builtins').diagnostics))
-	sources = vim.tbl_deep_extend('force', sources, vim.tbl_keys(require('null-ls.builtins').formatting))
-	sources = vim.tbl_deep_extend('force', sources, vim.tbl_keys(require('null-ls.builtins').code_actions))
-	sources = vim.tbl_deep_extend('force', sources, vim.tbl_keys(require('null-ls.builtins').completion))
-	sources = vim.tbl_deep_extend('force', sources, vim.tbl_keys(require('null-ls.builtins').hover))
-	local tools = lookup(removeArrayDuplicates(sources))
-	return tools
+	return _.filter_map(function(pkg_name)
+		return Optional.of_nilable(source_mapping.package_to_null_ls[pkg_name])
+	end, registry.get_installed_package_names())
 end
 
-local do_install = function(p, version, on_close)
-	if version ~= nil then
-		show(string.format('%s: updating to %s', p.name, version))
-	else
-		show(string.format('%s: installing', p.name))
+---@param filetype string | string[]
+local function is_source_in_filetype(filetype)
+	local filetype_mapping = require('mason-null-ls.mappings.filetype')
+
+	local function get_sources_by_filetype(ft)
+		return filetype_mapping[ft] or {}
 	end
-	p:once('install:success', function()
-		show(string.format('%s: successfully installed', p.name))
-	end)
-	p:once('install:failed', function()
-		show_error(string.format('%s: failed to install', p.name))
-	end)
-	p:install({ version = version }):once('closed', vim.schedule_wrap(on_close))
-end
 
-local check_install = function(force_update)
-	SETTINGS.ensure_installed =
-		vim.tbl_deep_extend('force', lookup(SETTINGS.null_ls_sources), SETTINGS.ensure_installed)
+	local source_candidates = _.compose(
+		_.set_of,
+		_.cond({
+			{ _.is('string'), get_sources_by_filetype },
+			{ _.is('table'), _.compose(_.flatten, _.map(get_sources_by_filetype)) },
+			{ _.T, _.always({}) },
```

Comment:
this is for my own knowledge. is this like an `else` block?

Reply:
Yes, everything that comes from `_` is logical tooling, mostly higher order functions to simplify list manipulation.